netann/chan_status_manager.go

```diff
@@ -0,0 +1,577 @@
+		case <-m.statusSampleTicker.C:
+			// First, do a sweep and mark any ChanStatusActive
+			// channels that are not active within the htlcswitch as
+			// ChanStatusPendingInactive. The channel will then be
+			// disabled if no request to enable is received before
+			// the ChanInactiveTimeout expires.
+			m.markPendingInactiveChannels()
+
+			// Now, do another sweep to disable any channels that
+			// were marked in a prior iteration as pending inactive
+			// if the inactive chan timeout has elapsed.
+			m.disableInactiveChannels()
+
+		case <-m.quit:
+			return
+		}
+	}
+}
+
+// processEnableRequest attempts to enable the given outpoint. If the method
+// returns nil, the status of the channel in chanStates will be
+// ChanStatusActive. If the channel is not active at the time of the request,
+// ErrEnableInactiveChan will be returned. An update will be broadcast only if
+// the channel is currently disabled, otherwise no update will be sent on the
+// network.
+func (m *ChanStatusManager) processEnableRequest(outpoint wire.OutPoint) error {
+	curState, err := m.getOrInitChanStatus(outpoint)
+	if err != nil {
+		return err
+	}
+
+	// Quickly check to see if the requested channel is active within the
+	// htlcswitch and return an error if it isn't.
+	chanID := lnwire.NewChanIDFromOutPoint(&outpoint)
+	if !m.cfg.IsChannelActive(chanID) {
+		return ErrEnableInactiveChan
+	}
+
+	switch curState.Status {
+
+	// Channel is already enabled, nothing to do. We can ignore whether the
+	// channel is actually enabled within the switch for this case, as it
+	// will be marked as pending inactive on the next sweep if it is still
+	// not active.
+	case ChanStatusActive:
+		return nil
+
+	// The channel is enabled, though we are now canceling the scheduled
+	// disable.
+	case ChanStatusPendingInactive:
+		log.Debugf("Channel(%v) already enabled, canceling scheduled "+
+			"disable", outpoint)
+
+	// We'll sign a new update if the channel is still disabled.
+	case ChanStatusInactive:
+		log.Infof("Announcing channel(%v) enabled", outpoint)
+
+		err := m.signNextUpdate(outpoint, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	m.chanStates.markActive(outpoint)
+
+	return nil
+}
+
+// processDisableRequest attempts to disable the given outpoint. If the method
+// returns nil, the status of the channel in chanStates will be
+// ChanStatusInactive. An update will only be sent if the channel has a status
+// other than ChanStatusActive, otherwise no update will be sent on the network.
+func (m *ChanStatusManager) processDisableRequest(outpoint wire.OutPoint) error {
+	curState, err := m.getOrInitChanStatus(outpoint)
+	if err != nil {
+		return err
+	}
+
+	switch curState.Status {
+
+	// Channel is already disabled, nothing to do.
+	case ChanStatusInactive:
+		return nil
+
+	// We'll sign a new update disabling the channel if the current status
+	// is enabled or pending-inactive.
+	case ChanStatusActive, ChanStatusPendingInactive:
```

Comment:
> Otherwise, we'll go ahead and sign a new update and also reset the SendDisabledTime for this channel. 

Not sure I follow, this is intentional because channels in `ChanStatusPendingInactive` are enabled on the network, so we want to immediately disable them and then forgo the passive disable.

> Ideally the caller doesn't repeatedly call RequestDisable for the same channel, but would be nice to not delegate that responsibility.

If the disable is requested multiple times, the subsequent ones will be ignored as they end up in the `ChanStatusInactive`